File: gpib_ethernet.py
"""
Try to connect to GPIB-ETHERNET (the adapter)
Performed on 2022 Mar 15
"""

import socket as skt
import time
import math
import os
# import numpy as np
# import numpy.lib.recfunctions as rfn
# import matplotlib
# import matplotlib.pyplot as plt

HOST = '192.168.1.32'
PORT = 1234

def connect():

    s = skt.socket()
    s.settimeout(5)
    try:
        s.connect((HOST,PORT))
        s.send(b'++savecfg 0\n')
        s.send(b'++addr 28\n')
        s.send(b'++auto 0\n')
        s.send(b'++ver\n')
        try:
            print(s.recv(1024))
            s.send(b'*IDN?\n')
            s.send(b'++read EOI\n')
            try:
                print(s.recv(1024))
                s.send(b'RF?\n')
                s.send(b'++read EOI\n')
                try:
                    print(s.recv(1024))
                    s.send(b'LEVEL?\n')
                    s.send(b'++read EOI\n')
                    try:
                        print(s.recv(1024))
                        s.send(b'LEVEL:OFF\n')
                        s.send(b'RF:START 10KHZ\n')
                        s.send(b'RF:STOP 62MHZ\n')
                        s.send(b'RF:STEP 200KHZ\n')
                        s.send(b'TIME:RF_SWP 50MS\n')
                        s.send(b'SWP:AUTO\n')
                        s.send(b'LEVEL:ON\n')
                    except skt.timeout:
                        print('Reading RF Output LEVEL Failed!')
                except skt.timeout:
                    print('Inquiring RF parameter Failed!')
            except skt.timeout:
                print('Reading GPIB Instrument Version Failed!')
        except skt.timeout:
            print('Reading GPIB-ETHERNET Module Version Failed!')
    except skt.timeout:
        print('GPIB-ETHERNET Module Network Error!')

    # rm = pyvisa.ResourceManager()
    # print('List GPIB Instruments:')
    # instru_tuple = rm.list_resources()
    # print('\t' + str(instru_tuple))
    # for idx in range(len(instru_tuple)):
        # try:
            # instru = rm.open_resource(str(instru_tuple[idx]))
            # instru_str = instru.query('*IDN?')
            # print('Instrument IDN of ' + str(instru_tuple[idx]) + ' is ' + instru_str)
        # except:
            # print('Error found when inquiring ' + str(instru_tuple[idx]))

def main():
    connect()
        

if __name__ == '__main__':
    main()

File: gpib_ethernet.c
#include<stdio.h>
#include<sys/socket.h>
#include<arpa/inet.h>
#include<string.h>

int gpib_config()
{
	int socket_desc;
	struct sockaddr_in server;
	char *server_addr = "192.168.1.32";
	char *message , server_reply[2000];

	//Create socket
	socket_desc = socket(AF_INET , SOCK_STREAM , 0);
	server.sin_addr.s_addr = inet_addr(server_addr);
	server.sin_family = AF_INET;
	server.sin_port = htons( 1234 );

	//Connect to remote server
	printf("Trying to connect to %s\n", server_addr);
	if (connect(socket_desc , (struct sockaddr *)&server , sizeof(server)) < 0)
	{
		puts("GPIB-ETHERNET Module Network Error!");
		return 1;
	}

	//Connect GPIB-Ethernet Module
	message = "++savecfg 0\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++addr 28\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++auto 0\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++ver\n";
	send(socket_desc , message , strlen(message) , 0);
	if( recv(socket_desc, server_reply , 1024 , 0) < 0)
	{
		puts("Reading GPIB-ETHERNET Module Version Failed!");
		return 1;
	}
	puts(server_reply);

	//Request GPIB Instrument Version
	message = "*IDN?\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++read EOI\n";
	send(socket_desc, message, strlen(message), 0);
	if( recv(socket_desc, server_reply , 1024 , 0) < 0)
	{
		puts("Reading GPIB Instrument Version Failed!");
		return 1;
	}
	puts(message);
	puts(server_reply);

	//Request RF Frequency
	message = "RF?\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++read EOI\n";
	send(socket_desc, message, strlen(message), 0);
	if( recv(socket_desc, server_reply , 1024 , 0) < 0)
	{
		puts("Inquiring RF parameter Failed!");
		return 1;
	}
	puts(server_reply);

	//Request RF Level
	message = "LEVEL?\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "++read EOI\n";
	send(socket_desc, message, strlen(message), 0);
	if( recv(socket_desc, server_reply , 1024 , 0) < 0)
	{
		puts("Reading RF Output LEVEL Failed!");
		return 1;
	}
	puts(server_reply);

	//Set RF Sweep
	message = "LEVEL:OFF\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "RF:START 10KHZ\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "RF:STOP 150MHZ\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "RF:STEP 400KHZ\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "TIME:RF_SWP 40MS\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "LEVEL:ON\n";
	send(socket_desc , message , strlen(message) , 0);
	message = "SWP:SINGLE\n";
	send(socket_desc , message , strlen(message) , 0);

	return 0;
}

int main(int argc , char *argv[])
{
	if (gpib_config())
	{
		puts("Exit with Error.");
	}
}
